File: algorithms/python/rotateList/rotateList.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/rotate-list/
# Author: Miao Zhang
# Date:   2021-01-13

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rotateRight(self, head: ListNode, k: int) -> ListNode:
        if head is None: return head;
        dummy = ListNode(0)
        dummy.next = head
        count = 0
        pi = dummy
        while pi.next:
            count += 1
            pi = pi.next
        k = k % count
        # still original list
        if k == 0 or count == 1: return head
        
        pi.next = head
        pi = dummy
        for _ in range(count - k):
            pi = pi.next
        pj = pi.next
        pi.next = None
        return pj

File: algorithms/cpp/rotateList/rotateList.cpp
// Source: https://leetcode.com/problems/rotate-list/
// Author: Miao Zhang
// Date:   2021-01-13

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if (head == nullptr) return nullptr;
        ListNode* dummy = new ListNode(0);
        dummy->next = head;
        int count = 0;
        ListNode* pi = dummy;
        while (pi->next) {
            count++;
            pi = pi->next;
        }
        k = k % count;
        // rotate k times original list
        if (k == 0 || count == 1) return head;
    
        pi->next = head;
        pi = dummy;
        for (int i = 0; i < count - k; i++) {
            pi = pi->next;
        }
        ListNode* pj = pi->next;
        pi->next = nullptr;
        return pj;
    }
};
